base/compiler/ssair/slot2ssa.jl

```diff
@@ -741,8 +741,8 @@ function construct_ssa!(ci::CodeInfo, ir::IRCode, domtree::DomTree,
                 unode = ivalundef ? UpsilonNode() : UpsilonNode(ival)
                 typ = ivalundef ? MaybeUndef(Union{}) : typ_for_val(ival, ci, ir.sptypes, -1, slottypes)
                 push!(node.values,
-                    NewSSAValue(insert_node!(ir, first_insert_for_bb(code, cfg, item),
-                                 NewInstruction(unode, typ), true).id - length(ir.stmts)))
+                    NewSSAValue(insert_node!(ir, first_insert_for_bb(code, cfg, item+1),
```

Comment:
sigh... This change isn't right, but neither is the one before. `:enter` needs to stay a BB terminator, but as the tests point out, we also can't just insert to the next block:
```
julia> @code_typed foo28224()
CodeInfo(
1 ─      nothing::Nothing
2 ─ %2 = $(Expr(:enter, #4))
3 ┄ %3 = φ (#2 => 0, #3 => %6)::Int64
│   %4 = ϒ (0)::Int64
│        invoke Main.make_error28224(%3::Int64)::Bool
│   %6 = Base.add_int(%3, 1)::Int64
│   %7 = ϒ (%6)::Int64
└──      goto #3
4 ┄ %9 = φᶜ (%4, %7)::Int64
└──      $(Expr(:leave, 1))
5 ─      $(Expr(:pop_exception, :(%2)))::Any
└──      return %9
) => Int64
```

Reply:
Hmm, so as far as I understand the purpose of this change is to reduce the number of basic blocks produced by `:enter` expression right? And I'm not sure if there is any valid way to pack those two upsilon nodes in a single block given their memory semantics. Maybe we can change the semantics of `:enter` a bit so that it can have succeeding statements...?

Reply:
The issue is that before we were trying to insert the UpsilonNode after the `:enter` expression, which is illegal, because `:enter` is a terminator. We didn't used to catch that, but I added a verifier check for it in this PR, because it was causing miscompiles in cfg_simplify, which looks at terminators. I think there's two possible ways two fix it. One is to split the edge after the `:enter` to have a place to put the upsilon node. The other is to put the upsilon node before the `:enter` (which I think is not allowed at the moment, but should be supportable from a codegen perspective).

Reply:
It should be okay to put them immediately before the enter (the inverse of the rule for phinodes being immediately after). That would let us still compute lifetimes (if we ever want to compact their stack usage). Though may be more annoying to compute the start of the block then (which would be the first upsilon before an enter that belongs to that enter—merging being okay)

Reply:
I guess this issue is now resolved (it seems like we don't need changes in codegen).